packages/itwin/tree-widget/src/components/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.ts

```diff
@@ -1048,3 +1024,35 @@ function setIntersection<T>(lhs: Set<T>, rhs: Set<T>): Set<T> {
   lhs.forEach((x) => rhs.has(x) && result.add(x));
   return result;
 }
+
+function reduceFilterPaths(paths: HierarchyNodeIdentifiersPath[]) {
+  const sorted = [...paths].sort((a, b) => a.length - b.length);
+  paths = [];
+  for (const path of sorted) {
+    if (!arraySortedByLengthContainsPrefix(paths, path)) {
+      paths.push(path);
+    }
+  }
+  return paths;
+}
+
+function arraySortedByLengthContainsPrefix<T>(targetArray: T[][], source: T[]) {
+  for (const targetVal of targetArray) {
+    if (targetVal.length >= source.length) {
+      break;
+    }
+
+    let isPrefix = true;
+    for (let i = 0; i < targetVal.length; ++i) {
+      if (targetVal[i] !== source[i]) {
```

Comment:
Changed to `equal`, although in this particular case regular `===` check would suffice because path elements references are shared between the arrays, but I admit, this is bug-prone